KAFKA-14412: Use shared cache for Task offset sums

nicktelford opened the pull request:
Instead of reading Task state offsets for non-open Tasks from the
`.checkpoint` file, we now maintain an in-memory cache of the latest
changelog offsets for every Task on the instance.

On start-up, this cache is seeded with the changelog offsets for every
on-disk StateStore. Running Active and Standby Tasks then update this
cache on every checkpoint to ensure it always reflects the offsets
on-disk.

This breaks the tight coupling between `TaskManager` and `.checkpoint`
files, which will enable us to remove `.checkpoint` files in a later
commit as part of KIP-1035.

Reviewers: Eduwer Camacaro<eduwerc@gmail.com>, Bill Bejeck
 <bbejeck@apache.org>

nicktelford committed:
KAFKA-14412: Use shared cache for Task offset sums

Instead of reading Task state offsets for non-open Tasks from the
`.checkpoint` file, we now maintain an in-memory cache of the latest
changelog offsets for every Task on the instance.

On start-up, this cache is seeded with the changelog offsets for every
on-disk StateStore. Running Active and Standby Tasks then update this
cache on every checkpoint to ensure it always reflects the offsets
on-disk.

This breaks the tight coupling between `TaskManager` and `.checkpoint`
files, which will enable us to remove `.checkpoint` files in a later
commit as part of KIP-1035.

nicktelford commented:
@ableegoldman @mjsax @cadonna I had some time to pick this up, so I've rebased it against `trunk` and re-submitted it.

eduwercamacaro commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:
```diff
@@ -309,6 +311,46 @@ private void closeStartupTasks(final Predicate<Task> predicate) {
         }
     }
 
+    public Map<TaskId, Long> taskOffsetSums(final Set<TaskId> tasks) {
```
It would be nice to add a unit test that verifies the output of this method after updating the task offsets. 

Also, I think this method should be considered the source of truth for the current lag of each taskId, so I would suggest modifying this method to return an unmodifiable map instead of a modifiable one. However, this will break the TaskManager when it tries to put a `Task.LATEST_OFFSET` for the tasks that didn't have a cached offset, but we can do that on the scope of this method. 

In other words, it is making this method to return an offset (either a valid one or Task.LATEST_OFFSET) for every single TaskId received in the arguments.

nicktelford replied:
> However, this will break the TaskManager when it tries to **put a Task.LATEST_OFFSET for the tasks that didn't have a cached offset**

Not sure what you're getting at here. If tasks don't have an offset cached in `StateDirectory`, it's because no local state exists on-disk (or if there is, that local state is corrupt and will be wiped).

`TaskManager` simply overwrites the cached offset for running active tasks with `LATEST_OFFSET`, because we know we have the very latest offset for these tasks. `StateDirectory` can't do this, because it doesn't know anything about the assignment, only the state on the local disk.

A simple solution for the mutability would be to just have `TaskManager` construct a new map, copying the offsets from the `StateDirectory#taskOffsetSums` result and adding the `LATEST_OFFSET`s as necessary. But is it really worth it?

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:
```diff
@@ -309,6 +311,46 @@ private void closeStartupTasks(final Predicate<Task> predicate) {
+    }
+
+    public void updateTaskOffsets(final TaskId taskId, final Map<TopicPartition, Long> changelogOffsets) {
+        if (changelogOffsets.isEmpty()) {
```
maybe flip this to `!changelogOffsets.isEmpty()` and move the `taskOffsetSums.put` in the body of the `if` statement - IMHO it's easier to read

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:
```diff
@@ -309,6 +311,46 @@ private void closeStartupTasks(final Predicate<Task> predicate) {
+                            taskId);
+                }
+                offsetSum += offset;
+                if (offsetSum < 0) {
```
Is this even possible given above we throw if an offset is less than zero?

nicktelford replied:
This whole `sumOfChangelogOffsets` method was transplanted over from `TaskManager`, and this logic has remained unchanged, so I can't speak to the reasoning for it.

My guess is that if `offset` is somehow negative, then `offsetSum` could reduce below `0`.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java:
```diff
@@ -596,6 +639,7 @@ private void cleanRemovedTasksCalledByCleanerThread(final long cleanupDelayMs) {
                         final long now = time.milliseconds();
                         final long lastModifiedMs = taskDir.file().lastModified();
                         if (now - cleanupDelayMs > lastModifiedMs) {
+                            removeTaskOffsets(id);
```
maybe I'm missing something but on line 561 the code calls `taskOffsetSums.clear()` but executes `removeTaskOffsets(id)` but the map should be empty at this point?

nicktelford replied:
I don't follow I'm afraid. Can you elaborate?

`taskOffsetSums.clear()` is in `close()`, but this `removeTaskOffsets(id)` is in `cleanRemovedTasksCalledByCleanerThread()`, which isn't called by `close()` or vice-versa?

bbejeck reviewed (commented):
Thanks for the PR @nicktelford, I made a pass over the non-testing code

bbejeck commented:
@nicktelford can you rebase this? After that we can get this merged soon.

nicktelford commented:
@bbejeck I've rebased and resolved the conflict. All the tests pass. I've taken a quick pass over your review comments, but I haven't had a chance to tackle @eduwercamacaro's suggestion yet. That's a Monday job :smile:

nicktelford committed:
Simplify method

bbejeck commented:
@nicktelford gentle nudge

nicktelford commented:
@bbejeck I've rebased against trunk, which now includes @eduwercamacaro's changes to startup task/stores. All tests pass on my machine (including integration tests).

Is there anything outstanding that I've missed?

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1242,35 +1239,20 @@ public void signalResume() {
+        // overlay latest offsets from assigned tasks
+        for (final Task task : tasks.values()) {
+            if (task.isActive() && task.state() == State.RUNNING && taskOffsetSums.put(task.id(), Task.LATEST_OFFSET) == null) {
+                log.error("Could not find cached offset for assigned ACTIVE Task {}", task.id());
```
We're logging this as an error here is there potential issues or negative impact?  Since this is an error should we throw here? Otherwise if it's not a problem maybe change to a warn?  Either way it might be good to briefly add more information so users don't get uncessarily concerned

nicktelford replied:
Hmm, I'm trying to think of situations where this might occur. This should only happen if we've been assigned an ACTIVE task with no previous local state, so in that case it's actually a valid corner case, the first time a task is assigned to an instance (e.g during fresh application startup). Since there are actually valid cases, perhaps we shouldn't log anything at all?

And looking at this in more detail, I think there's a bug here: this method is intended to exclude stateless and non-logged tasks, but this new code would include all tasks. I'll fix that now!

nicktelford replied:
@bbejeck I've removed this log message, fixed the bug, added a test for it and rebased against trunk, as requested.

bbejeck reviewed (commented):
Thanks @nicktelford ! Overall the lgtm I have 1 minor comment.  Also could we add a test for `StateDirectory.taskOffsetSums`

bbejeck commented:
@nicktelford can you rebase one more time? We merged some minor changes today we want picked up.

nicktelford committed:
Fix taskOffsetSums() including stateless tasks

When computing the taskOffsetSums() active and running tasks, we should
exclude both stateless and non-logged stateful tasks, as per the
original implementation of this method.

A new test has been added to validate this constraint.

bbejeck reviewed (approved):
Thanks for the updates @nicktelford, LGTM

bbejeck commented:
Merged #20954 into trunk
